Web Extensions: Get getTree() to display tree structure with mock nodes

akutira-umich opened the pull request:
#### 883a1e2a1a7a8d1d9a89e141e5d678f00fa54981
<pre>
Web Extensions: Get getTree() to display tree structure with mock nodes
<a href="https://rdar.apple.com/154437204">rdar://154437204</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295067">https://bugs.webkit.org/show_bug.cgi?id=295067</a>

Reviewed by Brian Weinstein.

Created root node and appended nodes to the parent&apos;s children array or root&apos;s
children array. Then returned the root. Created tests for these.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIBookmarks::initializeMockBookmarksInternal):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
(WebKit::WebExtensionAPIBookmarks::getTree):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateParse)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetTree)):

Canonical link: <a href="https://commits.webkit.org/297188@main">https://commits.webkit.org/297188@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c1aa009124d965abf25eaf506458b08dc7a0de0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110649 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20741 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116675 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60916 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112612 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30987 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38897 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84137 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113597 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24746 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99633 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64578 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24108 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17772 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60470 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94135 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17831 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119465 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37689 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27994 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93099 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/110146 "Build is in progress. Recent messages:") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38063 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95903 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92923 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15690 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33666 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17875 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37585 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43057 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37247 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40586 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38955 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -49,17 +49,61 @@ class WebExtensionAPIBookmarks : public WebExtensionAPIObject, public JSWebExten
         BookmarkTreeNodeType type;
+        Vector<Ref<MockBookmarkNode>> children;
+
+        // --- Explicit Copy Constructor ---
```
```suggestion
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
```diff
@@ -49,17 +49,61 @@ class WebExtensionAPIBookmarks : public WebExtensionAPIObject, public JSWebExten
+
+        }
+
+        // --- Explicit Copy Assignment Operator ---
```
```suggestion
```

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -43,8 +52,10 @@
+static NSString * const childrenKey = @"children";
 static NSString * const bookmarkKey = @"bookmark";
 static NSString * const folderKey = @"folder";
+static NSString * const BookmarksRootId = @"testroot";
```
variable should be lowercase.
`bookmarksRootId = @"testBookmarksRoot"`

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -63,7 +74,14 @@
 static NSDictionary *toWebAPI(const WebExtensionAPIBookmarks::MockBookmarkNode& node)
 {
-    auto tempNode = @{
+    NSMutableArray* childrenArray = nil;
```
```suggestion
    NSMutableArray* childrenArray;
```

kiaraarose replied:
`NSMutableArray *childrenArray = [NSMutableArray array];`
i think you can initialize it here, and then set the childrenArray below if this isn't empty

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -63,7 +74,14 @@
+            [childrenArray addObject:toWebAPI(child)];
+    }
+
+    NSDictionary* baseNodeDict = @{
```
`NSDictionary *baseNodeDict`. `*` should go next to the variable for ObjectiveC

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -73,6 +91,11 @@
         typeKey: toWebAPI(node.type, node.url.createNSString().get())
     };
 
+    NSMutableDictionary* tempNode = baseNodeDict.mutableCopy;
```
ditto

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -63,7 +74,14 @@
+        childrenArray = [NSMutableArray array];
+        for (const auto& child : node.children)
+            [childrenArray addObject:toWebAPI(child)];
+    }
```
Can this be done below after initializing `baseNodeDict`, so that we don't have to test `if (childrenArray)` on line 96? We can just have this conditional below, and have `[tempNode setObject:childrenArray forKey:childrenKey];` after the loop which populates `childrenArray`.

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -89,8 +112,49 @@
+    rootNode->type = WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder;
+    rootNode->dateAdded = WTF::WallTime::fromRawSeconds(0);
+    rootNode->index = 0;
+    rootNode->children.clear();
```
```suggestion
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -89,8 +112,49 @@
+        folderRef->type = WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder;
+        folderRef->dateAdded = WTF::WallTime::now();
+        folderRef->index = index;
+        folderRef->children.clear();
```
```suggestion
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -89,8 +112,49 @@
+    rootNode->children.clear();
+    m_mockBookmarks.add(rootNode->id, WTFMove(rootNode));
+
+    auto addStandardFolder = [&](const String& id, const String& title, int index) {
```
```suggestion
    auto addStandardFolder = [&](const String& identifier, const String& title, int index) {
```
`id` is a reserved keyword

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -89,8 +112,49 @@
+
+    auto addStandardFolder = [&](const String& id, const String& title, int index) {
+        Ref<MockBookmarkNode> folderRef = MockBookmarkNode::create();
+        folderRef->id = id;
```
```suggestion
        folderRef->id = identifier;
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -89,8 +112,49 @@
+        if (auto rootOptional = m_mockBookmarks.getOptional(BookmarksRootId)) {
+            if (auto folderOptional = m_mockBookmarks.getOptional(id))
+                (*rootOptional)->children.append(folderOptional.value());
+        }
```
```suggestion
        auto rootOptional = m_mockBookmarks.getOptional(BookmarksRootId);
        auto folderOptional = m_mockBookmarks.getOptional(id);
        if (rootOptional && folderOptional)
            rootOptional.value()->children.append(folderOptional.value());
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -121,19 +185,43 @@
+    newNode->id = bookmark[idKey];
+    newNode->url = bookmark[urlKey];
+    newNode->title = bookmark[titleKey];
+    newNode->parentId = bookmark[parentIdKey] ? bookmark[parentIdKey] : @"testroot";
```
```suggestion
    newNode->parentId = bookmark[parentIdKey] ? bookmark[parentIdKey] : bookmarksTestRootId;
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -121,19 +185,43 @@
+
+    m_mockBookmarks.add(newNodeId, newNode);
+
+    if (auto nodeInMapOptional = m_mockBookmarks.getOptional(newNodeId)) {
```
Unfortunately I can't put all of this in a suggested change due to github's limitations, but since you're using `Ref<MockBookmarkNode>`, you don't need to get the new node out of `m_mockBookmarks` after inserting it. `newNode` will still be a valid reference pointing to the same object as that pointed to by the reference in `m_mockBookmarks`, so you can just use `parentNode->children.append(newNode)` and then sort it, without getting anything back out of `m_mockBookmarks`.

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -121,19 +185,43 @@
+    Ref<MockBookmarkNode> parentNode = (parentOptional).value();
+
+    if (parentNode->type != WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder) {
+        *outExceptionString = toErrorString(nullString(), typeKey, @"it better be a folder'").createNSString().autorelease();
```
This error is indicating that the parent specified by the new node is not a folder, right? If so:
```suggestion
        *outExceptionString = toErrorString(nullString(), parentIdKey, @"it must specify a node which is a folder").createNSString().autorelease();
```

akutira-umich replied:
Yes that's correct

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -121,19 +185,43 @@
+    String newNodeId = newNode->id;
+
+    auto parentOptional = m_mockBookmarks.getOptional(parentId);
+    Ref<MockBookmarkNode> parentNode = (parentOptional).value();
```
```suggestion
    if (!parentNode) {
        // indicate an error here, since the parentId could not be mapped to a node
        return;
    }
    Ref<MockBookmarkNode> parentNode = parentOptional.value();
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -148,15 +236,28 @@
+            [resultArray addObject:toWebAPI(nodeRef)];
+            bookmarksFound++;
+        }
+    }
```
```suggestion
    for (const auto& nodeRef : allBookmarks) {
        if (resultArray.count >= numberOfItems)
            break;

        if (nodeRef->type == WebExtensionAPIBookmarks::BookmarkTreeNodeType::Bookmark)
            [resultArray addObject:toWebAPI(nodeRef)];
    }
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -167,7 +268,21 @@
+    if (it == m_mockBookmarks.end()) {
+        *outExceptionString = toErrorString(nullString(), @"root", @"root not found").createNSString().autorelease();
+        return;
+    }
```
Nit: my preference would be to use `getOptional` here rather than `find` to match the convention of the rest of the file.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -121,19 +182,40 @@
-    newNode.parentId = bookmark[parentIdKey];
-    newNode.index = parsedDetails.index.value();
-    newNode.type = parsedDetails.type.value_or(WebExtensionAPIBookmarks::BookmarkTreeNodeType::Folder);
+    Ref<MockBookmarkNode> newNode = MockBookmarkNode::create();
```
```suggestion
    Ref newNode = MockBookmarkNode::create();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -89,8 +109,49 @@
 
+void WebExtensionAPIBookmarks::initializeMockBookmarksInternal()
+{
+    Ref<MockBookmarkNode> rootNode = MockBookmarkNode::create();
```
```suggestion
    Ref rootNode = MockBookmarkNode::create();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -121,19 +182,40 @@
+        *outExceptionString = toErrorString(nullString(), parentIdKey, @"parentId could not be mapped to a node").createNSString().autorelease();
+        return;
+    }
+    Ref<MockBookmarkNode> parentNode = parentOptional.value();
```
```suggestion
    }

    Ref parentNode = parentOptional.value();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -121,19 +182,40 @@
+
+    auto parentOptional = m_mockBookmarks.getOptional(parentId);
+    if (!parentOptional) {
+        *outExceptionString = toErrorString(nullString(), parentIdKey, @"parentId could not be mapped to a node").createNSString().autorelease();
```
```suggestion
        *outExceptionString = toErrorString(nullString(), parentIdKey, @"it could not be mapped to a node").createNSString().autorelease();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -167,7 +258,22 @@
+    auto rootNodeOptional = m_mockBookmarks.getOptional(bookmarksRootId);
+    if (!rootNodeOptional) {
+        if (outExceptionString)
+            *outExceptionString = toErrorString(nullString(), @"root", @"root not found").createNSString().autorelease();
```
```suggestion
            *outExceptionString = toErrorString(nullString(), nullString(), @"root not found").createNSString().autorelease();
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -31,9 +31,18 @@
+#include <wtf/FastMalloc.h>
+#include <wtf/HashMap.h>
+#include <wtf/HashSet.h>
+#include <wtf/RefPtr.h>
```
None of these should be needed.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -31,9 +31,18 @@
 #if ENABLE(WK_WEB_EXTENSIONS_BOOKMARKS)
 
+#include <functional>
+#include <wtf/text/WTFString.h>
```
Shouldn't be needed.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -43,8 +43,10 @@
+static NSString * const childrenKey = @"children";
 static NSString * const bookmarkKey = @"bookmark";
 static NSString * const folderKey = @"folder";
+static NSString * const bookmarksRootId = @"testBookmarksRoot";
```
Is this a temporary constant? Or will it be changed later?

 Can it have a blank line above it with a comment indicating that these should be removed when we're not mocking nodes anymore?

akutira-umich replied:
So bookmarkRootId will be constant, but it being set to "testBookmarksRoot" will most likely be changed. I will add a comment about this

b-weinstein replied:
If the constant will still be around, I think this would be more clear as a `FIXME`. Sorry for the confusion! Something like:

`FIXME: (bug number) rename this when we aren't mocking bookmarks anymore.`

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -73,6 +75,15 @@
         typeKey: toWebAPI(node.type, node.url.createNSString().get())
     };
 
+    NSMutableDictionary *tempNode = baseNodeDict.mutableCopy;
```
I've never really seen us use `.mutableCopy`, but it looks like we do!

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -63,7 +65,7 @@
 static NSDictionary *toWebAPI(const WebExtensionAPIBookmarks::MockBookmarkNode& node)
 {
-    auto tempNode = @{
+    NSDictionary *baseNodeDict = @{
```
Nit: I'd spell out `dictionary` here.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -89,8 +100,49 @@
+            rootOptional.value()->children.append(folderOptional.value());
+    };
+
+    addStandardFolder(@"testfav", @"favorite", 0);
```
Nit: These names could be a bit more readable. `addStandardFolder(@"testFavorites", @"Favorites", 0)`.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -89,8 +100,49 @@
 void WebExtensionAPIBookmarks::createBookmark(NSDictionary *bookmark, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
 {
+    if (m_mockBookmarks.isEmpty())
+        initializeMockBookmarksInternal();
```
Nit: I'd add a blank line after here.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -121,19 +173,40 @@
+    std::sort(parentNode->children.begin(), parentNode->children.end(),
+        [](const Ref<MockBookmarkNode>& a, const Ref<MockBookmarkNode>& b) {
+            return a->index < b->index;
+        });
```
Nit: This is a bit of extra work. When the original parent folder was created, it was sorted, right (in `initializeMockBookmarksInternal`)? Shouldn't we be able to just insert in order here, instead of append and then sort?

However, given that this is just for mock bookmark nodes, I don't think you need to actually make a change here.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -211,15 +211,49 @@
+    Util::loadAndRunExtension(bookmarkOnManifest, @{ @"background.js": Util::constructScript(script) }, bookmarkConfig);
+}
+
+TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIgetTree)
```
Nit: `BookmarksAPIGetTree`.

Or this could just be `TEST_F(WKWebExtensionAPIBookmarks, GetTree)`, since `BookmarksAPI` is implied by the `WKWebExtensionAPIBookmarks`.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -211,15 +211,49 @@
+        @"let folder1 = await browser.bookmarks.create({id: 'folder1', type: 'folder', title: 'Top Folder 1'});",
+        @"let bm2 = await browser.bookmarks.create({id: 'bm2', title: 'Child Bookmark 2', url: 'http://example.com/bm2', parentId: folder1.id});",
+        @"let bm3 = await browser.bookmarks.create({id: 'bm3', title: 'Top Bookmark 3', url: 'http://example.com/bm3'});",
+        @"let tree = await browser.bookmarks.getTree();",
```
Nit: I'd use `bookmark1` instead of `bm1`.

b-weinstein reviewed (changes requested):
This looks good! Some small feedback I'd like to see addressed.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -211,15 +211,49 @@
+        @"browser.test.assertEq('testBookmarksRoot', root.id, 'Root node ID should be root');",
+        @"browser.test.assertEq('folder', root.type, 'Root node type should be folder');",
+        @"browser.test.assertTrue(root.children.length >= 1, 'Root should have at least one child (default folder)');",
+        @"let foundbookmark1 = root.children.find(n => n.id === bookmark1.id);",
```
Nit: `foundBookmark1` and `foundBookmark2` further down in this file.

akutira-umich committed:
Web Extensions: Get getTree() to display tree structure with mock nodes
rdar://154437204
https://bugs.webkit.org/show_bug.cgi?id=295067

Reviewed by Brian Weinstein.

Created root node and appended nodes to the parent's children array or root's
children array. Then returned the root. Created tests for these.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIBookmarks::initializeMockBookmarksInternal):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getRecent):
(WebKit::WebExtensionAPIBookmarks::getTree):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIBookmarks.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateParse)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetTree)):

Canonical link: https://commits.webkit.org/297188@main
